packages/test-utils/src/reporter.ts

```diff
@@ -920,52 +920,108 @@ class ReplayReporter<TRecordingMetadata extends UnstructuredMetadata = Unstructu
+    result: UploadableTestResult<TRecordingMetadata>,
+    executionGroupId: string
+  ) {
+    const executions = result.executions[executionGroupId];
+    const latestExecution = last(executions);
+    assert(!!latestExecution, "Expected at least one execution in the list");
+
+    let toUpload: typeof executions = [];
 
-    let toUpload: typeof results | undefined;
+    const aggregateStatus = this._assignAggregateStatus(result, executions);
 
-    switch (this._uploadStatusThreshold) {
-      case "all":
-        toUpload = this._minimizeUploads ? [latestResult] : results;
+    switch (aggregateStatus) {
+      case "failed": {
+        if (!this._minimizeUploads) {
+          result.didUploadStatuses.failed = true;
+          toUpload.push(...executions);
+          break;
+        }
+        if (result.didUploadStatuses.failed) {
+          break;
+        }
+        result.didUploadStatuses.failed = true;
+        toUpload.push(latestExecution);
         break;
-      case "failed-and-flaky":
-        // retries can be disables so we need to always check if the latest result is not passed
-        // with retries enabled we know that we only have to upload when there was more than one attempt at the test
-        // a single passed attempt can be safely ignored, multiple attempts mean that the test is flaky or failing
-        if (latestResult.result !== "passed" || results.length > 1) {
-          if (!this._minimizeUploads) {
-            toUpload = results;
-          } else if (results.length > 1) {
-            toUpload = [results[0], latestResult];
-          } else {
-            toUpload = [latestResult];
+      }
+      case "flaky": {
+        if (this._uploadStatusThreshold === "failed") {
+          break;
+        }
+        if (!this._minimizeUploads) {
+          result.didUploadStatuses.failed ||= executions.some(r => r.result !== "passed");
+          result.didUploadStatuses.passed ||= executions.some(r => r.result === "passed");
+
+          // currently previously completed execution groups that could be entirely passed or failed are not retroactively uploaded here
+          toUpload.push(...executions);
+
+          // fallthrough so we don't miss the other status when the flake was detected across different execution groups
+        }
+
+        if (!result.didUploadStatuses.failed) {
+          const failedExecution = Object.values(result.executions)
+            .flatMap(e => e)
+            .find(r => r.result !== "passed");
+
+          if (failedExecution) {
+            result.didUploadStatuses.failed = true;
+            toUpload.push(failedExecution);
+          }
+        }
+
+        if (!result.didUploadStatuses.passed) {
+          const passedExecution = Object.values(result.executions)
+            .flatMap(e => e)
+            .find(r => r.result === "passed");
+
+          if (passedExecution) {
+            result.didUploadStatuses.failed = true;
+            toUpload.push(passedExecution);
```

Comment:
The way this code is organized is a little hard for me to read but I think it looks correct.